common/common.go

```diff
@@ -695,3 +696,63 @@ func SetIfZero[T comparable](p *T, def T) bool {
 	*p = def
 	return true
 }
+
+var (
+	contextKeys   []any
+	contextKeysMu sync.RWMutex
+)
+
+// RegisterContextKey registers a key to be captured by FreezeCtx
+func RegisterContextKey(key any) {
+	contextKeysMu.Lock()
+	defer contextKeysMu.Unlock()
+	if !slices.Contains(contextKeys, key) {
+		contextKeys = append(contextKeys, key)
+	}
+}
+
+// FrozenContext holds captured context values
+type FrozenContext map[any]any
+
+// FreezeCtx captures values from the context for registered keys
+func FreezeCtx(ctx context.Context) FrozenContext {
+	contextKeysMu.RLock()
+	defer contextKeysMu.RUnlock()
+
+	values := make(FrozenContext, len(contextKeys))
+	for _, key := range contextKeys {
+		if val := ctx.Value(key); val != nil {
+			values[key] = val
+		}
+	}
+	return values
+}
+
+// ThawCtx creates a new context from the frozen context using context.Background() as parent
+func ThawCtx(fc FrozenContext) (context.Context, error) {
+	return MergeCtx(context.Background(), fc)
+}
+
+// MergeCtx adds the frozen values to an existing context
+func MergeCtx(ctx context.Context, fc FrozenContext) (context.Context, error) {
+	for k := range fc {
+		if ctx.Value(k) != nil {
+			return nil, fmt.Errorf("%w: %q", errDuplicateContextKey, k)
+		}
+	}
```

Comment:
I think this has to go. At scale, traversing a map and a context just incase there's existing values is likely to be a bottleneck. 

You are specifically asking to apply values from your frozen context, so I think it is okay just to apply them.

You may wish make it a new func entirely where you have `ApplyCtx` which doesn't error and doesn't loop instead and use `MergeCtx` for debugging

Reply:
Will remove the check